org.eclipse.lemminx/src/main/java/com/thaiopensource/relaxng/pattern/CMRelaxNGAttributeDeclaration.java

```diff
@@ -38,12 +42,48 @@ public class CMRelaxNGAttributeDeclaration implements CMAttributeDeclaration {
 
 	private final CMRelaxNGElementDeclaration cmElement;
 	private final AttributePattern pattern;
+
+	private String prefix;
+
+	private boolean computedPrefix;
+
 	private boolean required;
 	private List<String> values;
 
 	public CMRelaxNGAttributeDeclaration(CMRelaxNGElementDeclaration element, AttributePattern pattern) {
 		this.cmElement = element;
 		this.pattern = pattern;
+		this.computedPrefix = false;
+	}
+
+	@Override
+	public String getPrefix() {
+		String namespaceURI = getNamespace();
+		if (namespaceURI == null) {
+			return null;
+		}
+		if (computedPrefix) {
+			return prefix;
+		}
+		Locator locator = getPattern().getLocator();
+		if (locator != null) {
+			String documentURI = locator.getSystemId();
+			DOMDocument document = ((CMRelaxNGElementDeclaration) getOwnerElementDeclaration()).getCMDocument()
+					.getDocument(documentURI);
+			try {
+				int offset = document.offsetAt(new Position(locator.getLineNumber(), locator.getColumnNumber()));
+				DOMAttr attr = document.findAttrAt(offset);
+				if (attr != null) {
+					prefix = attr.getOwnerElement().getPrefix(namespaceURI);
+				} else {
+					prefix = document.getDocumentElement().getPrefix(namespaceURI);
+				}

```

Comment:
can any of those chained method calls (getPattern(), getOwnerElementDeclaration() ...) return something null?

Reply:
I clean my code. It should be better.